cql3/statements/alter_type_statement.cc

```diff
@@ -87,6 +87,21 @@ void alter_type_statement::do_announce_migration(database& db, ::keyspace& ks, b
         throw exceptions::invalid_request_exception(format("No user type named {} exists.", _name.to_string()));
     }
 
+    for (auto&& schema : ks.metadata()->cf_meta_data() | boost::adaptors::map_values) {
+        for (auto&& column : schema->partition_key_columns()) {
+            if (column.type->references_user_type(_name.get_keyspace(), _name.get_user_type_name())) {
+                throw exceptions::invalid_request_exception(format("Cannot add new field to type {} because it is used in the partition key of table {}",
+                    _name.to_string(), schema->cf_name()));
+            }
+        }
+        for (auto&& column : schema->clustering_key_columns()) {
+            if (column.type->references_user_type(_name.get_keyspace(), _name.get_user_type_name())) {
+                throw exceptions::invalid_request_exception(format("Cannot add new field to type {} because it is used in the clustering key of table {}",
+                    _name.to_string(), schema->cf_name()));
+            }
```

Comment:
These loops can cause a reactor stall in thousand-table schemas. Altering a user type is rare so it's not too bad.

Fixing it is hard. While it's easy to add yields in the loop, a concurrent update would invalidate iterators in ` ks.metadata()->cf_meta_data()` and segfault (if we're lucky). So a complete fix involves locking the entire schema, which is impossible now since multiple nodes/shards can alter tables concurrently. So this has to wait for raft-managed schema.